### PR TITLE
Esteban/feature/editable configs

### PR DESCRIPTION
**Pull request by esteban-gs**
##### Module for machete configs  (called machete settings in ui) within the Configs module
- Edit/read ui for machete user-configurable configs
  - JSON editor for multiple online hiring terms
- List ui for machete configs

image

image

### CODE REVIEW

File: src/app/configs/configs.service.ts [hunk 1]
```diff
@@ -50,7 +52,42 @@ export class ConfigsService {
   getConfig(key: string): Observable<Config> {
     return this.getAllConfigs().pipe(
       mergeMap((a) => a.filter((ll) => ll.key === key)),
+      map((a) => ({ ...a })), // return a new object
       first()
     );
   }
+
+  update(config: Config): Observable<Config> {
+    if (MS_NON_EDITABLE_CONFIGS_LOWER_CASE.includes(config.key.toLowerCase())) {
```
Comment (jcii) [inline src/app/configs/configs.service.ts:61]: Does the backend make this same check ? You have to assume the UI is untrustworthy and check on the server-side too.

File: src/app/configs/machete-settings/machete-settings-edit/Term.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Model for Machete Setting Terms
+ */
+export class Term {
```
Comment (jcii) [inline src/app/configs/machete-settings/machete-settings-edit/Term.ts:4]: We're using Term here and Confirm in the backend?

Comment (esteban-gs) [reply src/app/configs/machete-settings/machete-settings-edit/Term.ts]: Yes. I have updated it to be `OnlineOrderTerm`

File: src/app/configs/configs.service.ts (hunk 1 above)
Comment (esteban-gs) [reply src/app/configs/configs.service.ts:61]: I added a PR on the back end to validate

File: cypress/integration/settings/edit-settings.spec.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+          !MS_NON_EDITABLE_CONFIGS_LOWER_CASE.includes(c.key.toLowerCase())
+      );
+
+      privateConfigs = configs.filter(
+        (c: Config) =>
+          !c.publicConfig &&
+          MS_NON_EDITABLE_CONFIGS_LOWER_CASE.includes(c.key.toLowerCase())
+      );
+    });
+  });
```
Comment (esteban-gs) [inline cypress/integration/settings/edit-settings.spec.ts:41]: This is injecting test config data from `fixtures/settings`  into the http response

File: cypress/integration/settings/list-settings.spec.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    cy.getMacheteConfigs();
+    cy.visit(macheteSettingsRoutes.list);
+
+    cy.wait("@configs").then(() => {
+      configs = Cypress.env(ENV_KEY_MACHETE_CONFIGS) as Config[];
+      userEditableConfigs = configs.filter(
+        (c: Config) =>
+          c.publicConfig &&
+          !MS_NON_EDITABLE_CONFIGS_LOWER_CASE.includes(c.key.toLowerCase())
+      );
```
Comment (esteban-gs) [inline cypress/integration/settings/list-settings.spec.ts:34]: Also injecting test data in the http response

File: src/app/configs/configs-routing.module.ts [hunk 5]
```diff
@@ -19,6 +19,13 @@ const routes: Routes = [
       },
     ],
   },
+  {
+    path: "settings",
+    loadChildren: () =>
+      import("./machete-settings/machete-settings.module").then(
+        (m) => m.MacheteSettingsModule
+      ),
+  },
```
Comment (esteban-gs) [inline src/app/configs/configs-routing.module.ts:28]: this lazy-loads the module

File: src/app/configs/configs.service.ts [hunk 6]
```diff
@@ -50,7 +52,42 @@ export class ConfigsService {
...
+        pluck("data"),
+        map((data) => data as Config),
+        tap(() => {
+          this.appMessages.showSuccess({
+            label: "Success",
+            message: "Record Saved",
+          });
+        })
+      );
+  }
```
Comment (esteban-gs) [inline src/app/configs/configs.service.ts:92]: `catchError` will break the stream and user will receive the feedback

**Review by jcii — APPROVED** [review]

File: src/app/configs/machete-settings/machete-settings-edit/machete-settings-edit.component.html [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            <small class="p-error" *ngIf="!r.value"
+              >Value is required</small
+            >
+          </div>
+          <div class="p-field p-col-12" *ngIf="r.key === TERMS_KEY">
+            <app-machete-settings-term-form [termsAsString]="r.value" (termChange)="onChildTermChange($event, r)"></app-machete-settings-term-form>
+          </div>
+        </div>
+      </form>
+    </div>
```
Comment (esteban-gs) [inline src/app/configs/machete-settings/machete-settings-edit/machete-settings-edit.component.html:67]: logic to determine if form will display an input field or the `onlineTerms` form. The terms form is in a child component: `app-machete-settings-term-form`

File: src/app/configs/machete-settings/machete-settings-edit/machete-settings-edit.component.html [hunk 8]
```diff
@@ -0,0 +1,119 @@
+<ng-container *ngIf="record$ | async as r">
```
Comment (esteban-gs) [inline src/app/configs/machete-settings/machete-settings-edit/machete-settings-edit.component.html:1]: `foo$ | async as bar` in templates is a new ng feature that handles the subscribe/unsubscribe logic automatically

**Pull request merged**